File: packages/core/src/warehouseCatalog.ts
import type { RunoffDb } from "./db/index.js";
import { readWarehouseTables } from "./warehouse.js";
import type { CatalogFamily } from "./types/catalog.js";

/** Families → warehouse tables/columns/counts for one project. Server-only (root barrel, NOT /client-value-safe beyond types). */
export function buildWarehouseCatalog(db: RunoffDb, projectId: string): CatalogFamily[] {
  const fams = db.sqlite
    .prepare("SELECT id, key, label, kind, granularity FROM source_families WHERE project_id = ? ORDER BY key")
    .all(projectId) as Pick<CatalogFamily, "id" | "key" | "label" | "kind" | "granularity">[];
  const periodsStmt = db.sqlite.prepare(
    "SELECT period FROM sources WHERE family_id = ? AND status='filed' AND period IS NOT NULL ORDER BY period",
  );
  return fams.map((f) => {
    const tables = readWarehouseTables(projectId, f.key);
    return {
      ...f,
      queryable: tables.length > 0,
      tables,
      filedPeriods: f.kind === "constant" ? [] : (periodsStmt.all(f.id) as { period: string }[]).map((r) => r.period),
    };
  });
}

File: backend/runoff_api/core/warehouse_catalog.py
"""Port of packages/core/src/warehouseCatalog.ts."""

from runoff_api.core.db import RunoffDb
from runoff_api.core.warehouse import read_warehouse_tables


def build_warehouse_catalog(db: RunoffDb, project_id: str) -> list[dict]:
    """Families → warehouse tables/columns/counts for one project."""
    fams = db.execute(
        "SELECT id, key, label, kind, granularity FROM source_families WHERE project_id = ? ORDER BY key",
        (project_id,),
    ).fetchall()
    catalog: list[dict] = []
    for f in fams:
        tables = read_warehouse_tables(project_id, f["key"])
        if f["kind"] == "constant":
            filed_periods: list[str] = []
        else:
            filed_periods = [
                r["period"]
                for r in db.execute(
                    "SELECT period FROM sources WHERE family_id = ? AND status='filed' "
                    "AND period IS NOT NULL ORDER BY period",
                    (f["id"],),
                ).fetchall()
            ]
        catalog.append(
            {
                "id": f["id"],
                "key": f["key"],
                "label": f["label"],
                "kind": f["kind"],
                "granularity": f["granularity"],
                "queryable": len(tables) > 0,
                "tables": tables,
                "filedPeriods": filed_periods,
            }
        )
    return catalog
